### ISSUE TITLE
unable to build on ubuntu 19.10

### ISSUE THREAD

**Issue by sgqy** [Aegisub/Aegisub#158]
Version: 6f546951b4f004da16ce19ba638bf3eedefb9f31

## 1. command not found: `minilua`
Temporary fix: run `make` in `/vendor/luajit/src`

## 2. `Makefile.inc` error.
At line 139: `BIN_LUA      = ./vendor/luajit/src/host/minilua`
This also show command not found.
Temporary fix: `BIN_LUA      = $(PWD)/vendor/luajit/src/host/minilua`

## 3. Link error.
Command: 
```
g++ -o /home/aya/lab/Aegisub/automation/aegisub-lua  /home/aya/lab/Aegisub/automation/tests/aegisub.o /home/aya/lab/Aegisub/lib/libaegisub.a /home/aya/lab/Aegisub/lib/libluabins.a /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a -ldl -lm  -L/usr/lib/x86_64-linux-gnu -pthread   -lwx_gtk2u_gl-3.0 -lwx_gtk2u_stc-3.0 -lwx_gtk2u_xrc-3.0 -lwx_gtk2u_html-3.0 -lwx_gtk2u_qa-3.0 -lwx_gtk2u_adv-3.0 -lwx_gtk2u_core-3.0 -lwx_baseu_xml-3.0 -lwx_baseu_net-3.0 -lwx_baseu-3.0  -lz -L/usr/lib/x86_64-linux-gnu -lboost_filesystem -lboost_locale -lboost_regex -lboost_system -lboost_thread -lpthread -lboost_chrono -licuuc -licudata -licui18n -licuuc -licudata -pthread
```
Error:
```
/usr/bin/ld: /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a(ljamalg.o): in function `lj_err_throw':
/home/aya/lab/Aegisub/vendor/luajit/src/lj_err.c:452: multiple definition of `lj_err_throw'; /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a(lj_err.o):/home/aya/lab/Aegisub/vendor/luajit/src/lj_err.c:452: first defined here
/usr/bin/ld: /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a(ljamalg.o): in function `lua_atpanic':
/home/aya/lab/Aegisub/vendor/luajit/src/lj_err.c:758: multiple definition of `lua_atpanic'; /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a(lj_err.o):/home/aya/lab/Aegisub/vendor/luajit/src/lj_err.c:758: first defined here
/usr/bin/ld: /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a(ljamalg.o): in function `lj_func_closeuv':
/home/aya/lab/Aegisub/vendor/luajit/src/lj_func.c:81: multiple definition of `lj_func_closeuv'; /home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a(lj_func.o):/home/aya/lab/Aegisub/vendor/luajit/src/lj_func.c:81: first defined here
...
...(many)
```

## 4. Compile error.
Command:
```
g++ -MMD -MP -I/home/aya/lab/Aegisub/ -I/home/aya/lab/Aegisub/src/libresrc/ -I/usr/lib/x86_64-linux-gnu/wx/include/gtk2-unicode-3.0 -I/usr/include/wx-3.0 -D_FILE_OFFSET_BITS=64 -DWXUSINGDLL -D__WXGTK__ -Wall -Wextra -Wno-unused-parameter -fno-strict-aliasing -pipe -g -std=c++11 -Wno-c++11-narrowing -Wno-unused-local-typedefs -O3  -include acconf.h  -c -o /home/aya/lab/Aegisub/src/libresrc/libresrc.o /home/aya/lab/Aegisub/src/libresrc/libresrc.cpp
```
Error:
```
/home/aya/lab/Aegisub/src/libresrc/libresrc.cpp:26:42: error: no matching function for call to ‘wxBitmap::wxBitmap(wxImage, int, double&)’
   26 |   return wxBitmap(wxImage(mem), -1, scale);
      |                                          ^
/home/aya/lab/Aegisub/src/libresrc/libresrc.cpp:27:50: error: no matching function for call to ‘wxBitmap::wxBitmap(wxImage, int, double&)’
   27 |  return wxBitmap(wxImage(mem).Mirror(), -1, scale);
      |                                                  ^
```
Warning:
```
/usr/include/wx-3.0/wx/string.h:1224:25: warning: implicitly-declared ‘constexpr wxUniChar::wxUniChar(const wxUniChar&)’ is deprecated [-Wdeprecated-copy]
 1224 |     { assign(nRepeat, ch); }
      |                         ^
...
...
/usr/include/wx-3.0/wx/list.h:874:60: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
  874 |                 { return *(pointer_type)m_node->GetDataPtr(); }             \
      |                                                            ^

...
...(many)
```

## 5. Info.
`gcc -v`:
```
sing built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.2.1-9ubuntu2' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.2.1 20191008 (Ubuntu 9.2.1-9ubuntu2) 
```
`dpkg -l`:
```
libwxbase3.0-dev:amd64                         3.0.4+dfsg-12
libboost-all-dev                               1.67.0.2
g++                                            4:9.2.1-3.1ubuntu1
g++-9                                          9.2.1-9ubuntu2
```
`lsb_release -a`:
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 19.10
Release:	19.10
Codename:	eoan
```
`uname -a`:
```
Linux aya 4.19.79-041979-lowlatency #201910111332 SMP PREEMPT Fri Oct 11 17:39:51 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by wangqr** [comment]
As for 4, it is also seen in [travis](https://travis-ci.org/Aegisub/Aegisub/jobs/594417164). So you can try revert the last commit.

**Comment by wangqr** [comment]
For those lua related error, confirm you are running `make` in the root directory of this repo?

**Comment by sgqy** [comment]
> For those lua related error, confirm you are running `make` in the root directory of this repo?

I am running in root of repo, because `$(PWD)` works.

**Comment by wangqr** [comment]
That's weird, as it is not reproducible in Travis. The command to build aegisub is
```sh
./autogen.sh
./configure
make
```
If possible, can you provide the full log? You can use gist or pastebin if it is too long.

**Comment by sgqy** [comment]
> That's weird, as it is not reproducible in Travis. The command to build aegisub is
> 
> ```shell
> ./autogen.sh
> ./configure
> make
> ```
> 
> If possible, can you provide the full log? You can use gist or pastebin if it is too long.

https://pastebin.ubuntu.com/p/WSBQr3W2J7/

**Comment by wangqr** [comment]
I think I see what's going on here.
**tl;dr**: Install the system provided lua, run `git clean -fdx -- vendor`, and build again.

Explaination:
For 2, the `BIN_LUA` is wrong, and your fix is correct. System provided lua has a higher priority than bundled one, so it should hide the problem.
3 is actually caused by 1, because we should never built the normal version of lua. When using bundled luajit, aegisub [builds the amalg](https://github.com/Aegisub/Aegisub/blob/ce658d070925effea8c626b2ada2f819d01ab4fb/vendor/luajit/Makefile#L6) (means "everything in single large .o") version of luajit. If you manually built the normal version, they will confilct.

**Comment by sgqy** [comment]
> As for 4, it is also seen in [travis](https://travis-ci.org/Aegisub/Aegisub/jobs/594417164). So you can try revert the last commit.

I used second last version (85f711fccc75f01fd44f25537b8777df10c4b3d1). It compiles. But still link error.

Temporary fix:

1. Hide `vendor/luajit/src/ljamalg.c` and rebuild lua (use temporary solution: run `make` in `src` dir).
This fixed `aegisub-lua` link error (multiple definition).
2. Append `/home/aya/lab/Aegisub/vendor/luajit/src/libluajit.a` to the link list of `aegisub`.
This fixed `aegisub` link error.
The error message is:
```
/usr/bin/ld: /home/aya/lab/Aegisub/lib/libaegisub.a(modules.o): in function `agi::lua::preload_modules(lua_State*)':
/home/aya/lab/Aegisub/libaegisub/lua/modules.cpp:32: undefined reference to `luaL_openlibs'
collect2: error: ld returned 1 exit status
```

The program works, though.

**Comment by sgqy** [comment]
> I think I see what's going on here.
> **tl;dr**: Install the system provided lua, run `git clean -fdx -- vendor`, and build again.
> 
> Explaination:
> For 2, the `BIN_LUA` is wrong, and your fix is correct. System provided lua has a higher priority than bundled one, so it should hide the problem.
> 3 is actually caused by 1, because we should never built the normal version of lua. When using bundled luajit, aegisub [builds the amalg](https://github.com/Aegisub/Aegisub/blob/ce658d070925effea8c626b2ada2f819d01ab4fb/vendor/luajit/Makefile#L6) (means "everything in single large .o") version of luajit. If you manually built the normal version, they will confilct.

`apt install lua5.1` then restart everything.
Line 139 of `Makefile.inc` changed to `BIN_LUA      = lua`
(also installed `libluajit-5.1-dev` but do not know if it affects)

The whole progress works.

But `minilua` may still a problem...

**Comment by wangqr** [comment]
Note that 2 should be fixed with Aegisub/Aegisub#141
